Document Streaming API (#2133)
openvinotoolkit/model_server: docs/streaming_endpoints.md

> # Streaming API (preview) {#ovms_docs_streaming_endpoints}

Reviewer: ```suggestion
# gRPC Streaming API (preview) {#ovms_docs_streaming_endpoints}
```

## Introduction
OpenVINO Model Server implements gRPC KServe extension which adds separate RPC for [bidirectional streaming](https://grpc.io/docs/what-is-grpc/core-concepts/#bidirectional-streaming-rpc) use cases. It means that besides unary RPC, where client sends a single request and gets back a single response, client is able to initiate connection and send/receive messages in any number and order using `ModelStreamInfer` procedure:

```
service GRPCInferenceService
{
  ...
  // Standard
  rpc ModelInfer(ModelInferRequest) returns (ModelInferResponse) {}
  // Extension
  rpc ModelStreamInfer(stream ModelInferRequest) returns (stream ModelStreamInferResponse) {}
  ...
}
```

> This becomes very useful for endpoints serving [MediaPipe Graphs](./mediapipe.md). In unary inference RPC each request initializes new MediaPipe graph, inputs are deserialized into MediaPipe packets and pushed into the graph. After the graph is done, it is deleted.

Reviewer: ```suggestion
This becomes very useful for serving [MediaPipe Graphs](./mediapipe.md). In unary inference RPC, each request is using separate and independent MediaPipe graph instance.
```

> However, in streaming inference RPC MediaPipe graph is created only once per connection, and is reused by subsequent requests to the same gRPC stream. This removes graph initialization overhead and increases its overall throughput.

Reviewer: ```suggestion
However, in streaming inference RPC, MediaPipe graph is created only once per the client session and is reused by subsequent requests to the same gRPC stream. This enables handling stateful graphs where results depends on the previous state. It also adds support graphs with Mediapipe `source calculators` and enables special cases when the request timestamps can be defined by the client. Finally, it removes graph initialization overhead and increases its overall throughput.
```

Author: Timestamp and statefulness is explained below. I think it's not worth mentioning about everything in one paragraph @dtrawins

> However, in streaming inference RPC MediaPipe graph is created only once per connection, and is reused by subsequent requests to the same gRPC stream. This removes graph initialization overhead and increases its overall throughput.

Reviewer 2: ```suggestion
However, in streaming inference RPC MediaPipe graph is created only once per connection, and is reused by subsequent requests to the same gRPC stream. This avoids graph initialization overhead and increases its overall throughput.
```

![diagram](streaming_diagram.svg)

## Graph Selection

> Opening stream is not enough to create execution graph. Instead, the first gRPC request after the opening defines which graph definition will be selected for execution (`model_name` and `model_version` proto fields).

Reviewer: ```suggestion
After opening a stream, the first gRPC request defines which graph definition will be selected for execution (`model_name` proto fields).
```

> Afterwards, subsequent requests are required to match the servable name and version, otherwise the error is reported and input packets are not pushed to the graph. However, the graph remains in available for correct requests.

Reviewer: we don't version the graphs, why the version is mentioned. Empty version should be used.

Author: In current implementation, OVMS validates model name **and version** of subsequent requests. It must match first request model name **and version**.

Example:
First request - name: `graph_A`, version: `1`
Second request - name: `graph_A`, version: `500` <-- will be rejected

> Afterwards, subsequent requests are required to match the servable name and version, otherwise the error is reported and input packets are not pushed to the graph. However, the graph remains in available for correct requests.

Reviewer 2: ```suggestion
Afterwards, subsequent requests are required to match the servable name and version, otherwise the error is reported and input packets are not pushed to the graph. However, the graph remains available for correct requests.
```

> > NOTE: The server closes the stream after the first request if requested graph is non-existent or retired.

Reviewer: add info what happens if new requests have incorrect graph name and when in other cases the session is closed by the server

Author: There is information what happens, when subsequent (2+) request has incorrect name and version.

> Afterwards, subsequent requests are required to match the servable name and version...

Isn't it enough?

Reviewer: that is enoguh, in which other cases the session will be closed?

Author: The stream is closed when graph encounters an irrecoverable error. However, graph is probed for errors in Read loop, meaning - we only check for errors in-between subsequent client requests. Will add it to documentation.

Author: Added

## Timestamping
MediaPipe Graphs require packets to include timestamp information for synchronization purposes. Each input stream in the graph requires timestamps to be monotonically increasing. Read further about [MediaPipe timestamping](https://developers.google.com/mediapipe/framework/framework_concepts/synchronization#timestamp_synchronization).

### Automatic timestamping
By default OpenVINO Model Server assigns timestamps automatically. Each gRPC request is treated as separate point on the timeline, starting from 0. Each request is deserialized sequentially and increases the timestamp by 1.

> NOTE: It means that in order to send multiple inputs with the same timestamp, client needs to pack it into single request (or use manual timestamping).

### Manual timestamping
Optionally, the client is allowed to include timestamp manually via request parameter `OVMS_MP_TIMESTAMP`. It is applied to all the packets deserialized from the request.

It is possible to mix manual/automatic timestamping. After correct deserialization step, default automatic timestamp is always equal to `previous_timestamp + 1`.

> 

Reviewer: how the timestamp can be read from the response?

Author: Shouldn't it be included in the client snippet section? @bstrzele 
https://github.com/openvinotoolkit/model_server/blob/main/docs/clients_kfs.md

Reviewer: it should be in the snippets but we could mention that this value is also returned

Author: Added

## Preserving State Between Requests
Note that subsequent requests in a stream have access to the same instance of MediaPipe graph. It means that it is possible implement graph that saves intermediate state and act in stateful manner. It might be an advantage f.e. for object tracking use cases.

## Useful links
- example client snippets TODO

> - [complete demo with streaming](../demos/mediapipe/holistic_tracking/README.md)

Reviewer 2: Here we could use direct link to paragraph

Author: Will add once @bstrzele adds streaming variant

> > NOTE: Streaming API does not support requesting single models nor DAGs (Directed Acyclic Graphs) - those need to be included in MediaPipe Graph in order to use streaming.

Reviewer: ```suggestion
> NOTE: gRPC Streaming API supports only MediaPipe Graph servables via KServe API compatible client like `tritonclient`.
```

## Introduction
OpenVINO Model Server implements gRPC KServe extension which adds separate RPC for [bidirectional streaming](https://grpc.io/docs/what-is-grpc/core-concepts/#bidirectional-streaming-rpc) use cases. It means that besides unary RPC, where client sends a single request and gets back a single response, client is able to initiate connection and send/receive messages in any number and order using `ModelStreamInfer` procedure:

```
service GRPCInferenceService
{
  ...
  // Standard
  rpc ModelInfer(ModelInferRequest) returns (ModelInferResponse) {}
  // Extension
  rpc ModelStreamInfer(stream ModelInferRequest) returns (stream ModelStreamInferResponse) {}
  ...
}
```

This becomes very useful for serving [MediaPipe Graphs](./mediapipe.md). In unary inference RPC, each request is using separate and independent MediaPipe graph instance.
However, in streaming inference RPC MediaPipe graph is created only once per the client session, and is reused by subsequent requests to the same gRPC stream.

- This avoids graph initialization overhead and increases its overall throughput
- Allows to preserve state between subsequent incoming requests

> - Enables source calculators to stream multiple responses without feeding input packets

Reviewer 3: Hmm, can we put it in simpler words maybe? That point might be quite mysterious for someone who's not very familiar with mediapipe.

![diagram](streaming_diagram.svg)

## Graph Selection
After opening a stream, the first gRPC request defines which graph definition will be selected for execution (`model_name` proto field).
Afterwards, subsequent requests are required to match the servable name and version, otherwise the error is reported and input packets are not pushed to the graph. However, the graph remains available for correct requests.

> > **NOTE**: The server closes the stream after the first request if requested graph is non-existent or retired.

Reviewer 3: ```suggestion
> **NOTE**: The server closes the stream after the first request if requested graph does not exist or is retired.
```
